core/spec/models/spree/user_spec.rb

```diff
@@ -51,21 +51,36 @@
         expect(user.ship_address).to eq order.ship_address
       end
 
-      it "doesnt create new addresses if user has already" do
+      it "clones over order addresses" do
+        addresses_were_cloned
+      end
+
+      it "clones the addresses even if user has already" do
         user.update_column(:bill_address_id, create(:address))
         user.update_column(:ship_address_id, create(:address))
         user.reload
 
+        addresses_were_cloned
+      end
+
+      it "doesn't clone the address again if it is idential" do
+        bill_address_clone = order.bill_address.dup
+        ship_address_clone = order.ship_address.dup
+        bill_address_clone.save!; ship_address_clone.save!
+
+        user.update_column(:bill_address_id, bill_address_clone)
+        user.update_column(:ship_address_id, ship_address_clone)
+        user.reload
+
+        expect(user.bill_address).to eq order.bill_address
+        expect(user.ship_address).to eq order.ship_address
+
         expect {
           user.persist_order_address(order)
         }.not_to change { Spree::Address.count }
-      end
```

Comment:
Also note that this spec doesn't check that the user's addresses now equal the orders addresses, just that no new addresses were created. That's why the spec didn't catch our specific bug scenario.